addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java

```diff
@@ -0,0 +1,453 @@
+                outputStream.flush();
+            }
+        } catch (IOException e) {
+            String msg = "Exception writing to vallox: " + e.toString();
+            logger.error(msg);
+            for (StatusChangeListener l : this.getStatusListener()) {
+                l.statusChanged(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, msg);
+            }
+            throw e;
+        }
+    }
+
+    static byte calculateChecksum(byte[] pTelegram) {
+        int checksum = 0;
+        for (byte i = 0; i < pTelegram.length - 1; i++) {
+            checksum += pTelegram[i];
+        }
+        return (byte) (checksum % 256);
+    }
+
+    /**
+     * Start listening to telegrams on the serial interface. Stop if requested.
+     */
+    public void startListening() {
+        shutdownListening = false;
+        listenerExecutor = Executors.newSingleThreadExecutor();
+        listenerExecutor.submit(() -> {
+            while (!shutdownListening) {
+                try {
+                    Telegram telegram = receive();
+                    if (telegram == null) {
+                        // telegram too short or invalid
+                        continue;
+                    }
+                    logger.trace(telegram.toString());
+                    telegram.updateStore(vallox, vallueListener);
+                } catch (Exception e) {
+                    logger.error("Exception reading input stream: " + e.toString());
+                    try {
+                        Thread.sleep(2000);
+                    } catch (InterruptedException e1) {
+                        // ignore, simply retry, i.e. wait until reconnected
+                    }
+                }
+            }
+        });
+        logger.info("Start listening to vallox telegrams!");
+    }
+
+    /**
+     * Start a heartbeat thread that regularly polls a vallox variable and
+     * automatically reconnects if it fails.
+     */
+    public void startHeartbeat() {
+        heartbeatExecutor = Executors.newSingleThreadExecutor();
+        heartbeatExecutor.submit(() -> {
+            while (!shutdownListening) {
+                try {
+                    Thread.sleep(MINUTE);
+                    sendPoll(ValloxProperty.SelectStatus);
+                } catch (Exception e) {
+                    String msg = "Exception sending heartbeat poll: " + e.toString();
+                    logger.error(msg);
+                    for (StatusChangeListener l : this.getStatusListener()) {
+                        l.statusChanged(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, msg);
+                    }
+                    try {
+                        this.reconnect();
+                    } catch (IOException e1) {
+                        logger.error("Exception reconnecting vallox: " + e.toString());
+                    }
+                }
+            }
+        });
+        logger.info("Start heartbeat check to vallox!");
```

Comment:
Please change the logging level to `debug`. `trace` can be used for even more details, where necessary.